File: src/avenue_mcp/tools/classlist.py
"""get_class_list (docs/03-mcp-tools.md).

Written for the restricted case from the start. Full student rosters with
email addresses are personal information under FIPPA, and McMaster is more
likely to lock that down than less — a 403 is the expected and correct
outcome, not a bug to work around.

The route is under `le`, not `lp`. An earlier draft had it wrong, which
matters: the wrong path returns 404, and a 404 read as "blocked" would
permanently degrade this tool on the strength of a typo.
"""

from __future__ import annotations

import logging
from typing import Any

from avenue_mcp.client.d2l import D2LClient
from avenue_mcp.errors import AvenueMCPError

log = logging.getLogger(__name__)

# Role names that count as "staff" for the fallback. Brightspace lets
# institutions rename roles, so this is a heuristic over the role string
# rather than an ID lookup.
_STAFF_MARKERS = ("instructor", "teacher", "professor", "ta", "teaching assistant", "faculty")


def _is_staff(role: str) -> bool:
    lowered = (role or "").lower()
    return any(marker in lowered for marker in _STAFF_MARKERS)


def _person(entry: dict[str, Any]) -> dict[str, Any]:
    first = entry.get("FirstName") or ""
    last = entry.get("LastName") or ""
    name = (entry.get("DisplayName") or f"{first} {last}").strip()
    role = entry.get("RoleName") or (entry.get("Role") or {}).get("Name") or ""
    return {
        "name": name,
        "role": role,
        "email": entry.get("Email") or entry.get("EmailAddress") or None,
    }


async def get_class_list(client: D2LClient, *, org_unit_id: int) -> dict[str, Any]:
    """People associated with a course — instructors and TAs where available."""
    people: list[dict[str, Any]] = []
    roster_available = False
    note = ""

    for route in (f"{org_unit_id}/classlist/paged/", f"{org_unit_id}/classlist/"):
        try:
            raw = await client.get_json("le", route)
        except AvenueMCPError as exc:
            log.debug("classlist route %s failed: %s", route, exc)
            note = exc.message
            continue

        entries = raw if isinstance(raw, list) else (raw or {}).get("Items", [])
        people = [_person(e) for e in entries if isinstance(e, dict)]
        roster_available = True
        note = ""
        break

    if not roster_available:
        # Fallback: role-filtered enrollments. Often carries the same
        # restriction, and that's fine — instructor contact info is the part a
        # student actually needs ("who do I email about this?").
        try:
            entries = await client.get_paged(
                "lp", f"enrollments/orgUnits/{org_unit_id}/users/"
            )
            people = [
                _person(e.get("User") or e)
                for e in entries
                if isinstance(e, dict)
            ]
            note = "Full roster unavailable; showing enrollments visible to your account."
        except AvenueMCPError as exc:
            log.debug("role-filtered enrollments failed: %s", exc)
            note = (
                "Neither the class list nor the enrollment list is accessible from a "
                "student account on this Avenue instance. This is expected — student "
                "rosters are restricted personal information."
            )

    instructors = [p for p in people if _is_staff(p["role"])]
    students = [p for p in people if not _is_staff(p["role"])]

    if students and not note:
        # Withholding this is a deliberate choice, not a limitation, so say so
        # rather than letting an empty list read as "no classmates found".
        note = (
            f"{len(students)} student records were returned by Avenue but are not "
            f"included here. A full roster with names and emails is personal "
            f"information under FIPPA, and this tool returns instructor and TA "
            f"contacts by design."
        )

    return {
        "org_unit_id": org_unit_id,
        "instructors": instructors,
        "instructor_count": len(instructors),
        # The count answers "how big is this class" without handing over
        # a list of other people's names and email addresses.
        "student_count": len(students),
        "student_roster_returned": False,
        "note": note or None,
    }

File: extension/src/tools/classlist.ts
/**
 * get_class_list — course staff, and honesty about the roster.
 * Ported from src/avenue_mcp/tools/classlist.py.
 *
 * TWO THINGS HERE ARE LOAD-BEARING AND BOTH WERE LEARNED THE HARD WAY:
 *
 * 1. The route is under `le`, NOT `lp`. The wrong component returns 404, which
 *    the error path records as "roster unavailable" — degrading the tool
 *    permanently on a typo rather than a real permission boundary.
 *
 * 2. Roles arrive in `ClasslistRoleDisplayName`. Not `Role`, `RoleName`, or
 *    `RoleDisplayName` — none of which are present. Measured on this instance:
 *    "Instructor" x1, "TA 1" x10, "Student" x134. With that key missing every
 *    entry normalized to Unknown, all 145 landed in students, and `instructors`
 *    came back empty for a course with 11 staff.
 *
 * AND THE POLICY DECISION: the roster IS available here — 145 people with
 * names, McMaster emails, usernames, and student IDs. docs/02 predicted a 403
 * and the tool was written for that. Returning the array would ship a third of
 * a lecture hall's personal information to a model provider on every call, so
 * students are counted and withheld. docs/07 is unambiguous that this is FIPPA
 * data and that this tool exists to return instructor contacts.
 */

import { avenue } from "../avenue/client.js";
import { AvenueError } from "../avenue/errors.js";
import { isObj, normalizeRole, pick } from "../avenue/models.js";
import { courseName } from "./context.js";

interface Person {
  name: string | null;
  role: ReturnType<typeof normalizeRole>;
  email: string | null;
}

export async function getClassList(args: { org_unit_id: number }) {
  const { org_unit_id } = args;

  let people: Person[] = [];
  let rosterReadable = false;
  let note: string | null = null;

  try {
    const raw = await avenue.getPaged("le", `${org_unit_id}/classlist/`);
    people = raw.filter(isObj).map(toPerson);
    rosterReadable = true;
  } catch (err) {
    if (!(err instanceof AvenueError)) throw err;
    note =
      err.kind === "PermissionDenied"
        ? "The class list is not accessible from a student account on this course — this is expected, not an error."
        : `Could not read the class list: ${err.message}`;
  }

  const staff = people.filter((p) => p.role === "Instructor" || p.role === "TA");
  const students = people.filter((p) => p.role !== "Instructor" && p.role !== "TA");

  if (staff.length && !staff.some((p) => p.email)) {
    // Measured: the classlist populates DisplayName/FirstName/LastName/Username
    // but leaves Email empty for staff. Usernames are present and McMaster
    // addresses are conventionally <macid>@mcmaster.ca — but composing one
    // would be a guess presented as a contact detail.
    note =
      "Avenue did not provide email addresses for course staff. Names and roles " +
      "are accurate; find contact details on the course homepage or in the outline " +
      "rather than guessing an address.";
  }

  if (students.length && !note) {
    note =
      `${students.length} student records were returned by Avenue but are ` +
      `deliberately not included here. A roster with names and emails is personal ` +
      `information under FIPPA, and this tool returns course staff by design.`;
  }

  return {
    org_unit_id,
    course_name: await courseName(org_unit_id),
    instructors: staff,
    // Never returned in bulk, even when Avenue hands them over. The count
    // answers "how big is this class" without handing over the list.
    students: [] as Person[],
    student_roster_returned: false,
    student_count: rosterReadable ? students.length : null,
    note,
  };
}

function toPerson(entry: Record<string, unknown>): Person {
  const nested = pick(entry, "User");
  const user = isObj(nested) ? nested : entry;

  const first = String(pick(user, "FirstName", "GivenName") ?? "");
  const last = String(pick(user, "LastName", "Surname") ?? "");
  const display = String(pick(user, "DisplayName", "FullName") ?? `${first} ${last}`).trim();

  // ClasslistRoleDisplayName FIRST — see the header note. RoleId is an integer
  // and matches no hint, so it is the last resort rather than the first try.
  let role = normalizeRole(
    pick(entry, "ClasslistRoleDisplayName", "Role", "RoleName", "RoleDisplayName", "RoleAlias"),
  );
  if (role === "Unknown") role = normalizeRole(pick(entry, "RoleId"));

  return {
    name: display || null,
    role,
    email: (pick(user, "EmailAddress", "Email", "ExternalEmail") as string) ?? null,
  };
}
